include/PetriEngine/Colored/Reduction/RedRulePreemptiveFiring.h

```diff
@@ -0,0 +1,33 @@
+/*
+ * Authors:
+ *      Nicolaj Østerby Jensen
+ *      Jesper Adriaan van Diepen
+ *      Mathias Mehl Sørensen
+ */
+
+#ifndef VERIFYPN_REDRULEPREEMPTIVEFIRING_H
+#define VERIFYPN_REDRULEPREEMPTIVEFIRING_H
+
+#include "ReductionRule.h"
+#include "PetriEngine/Colored/ColoredNetStructures.h"
+
+namespace PetriEngine::Colored::Reduction {
+    class RedRulePreemptiveFiring : public ReductionRule {
+    public:
+        std::string name() override { return "PreemptiveFiring"; }
+
+        bool isApplicable(QueryType queryType, bool preserveLoops, bool preserveStutter) const override {
+            return !preserveStutter;
+        }
+
+        bool apply(ColoredReducer &red, const PetriEngine::PQL::ColoredUseVisitor &inQuery, QueryType queryType,
+                   bool preserveLoops, bool preserveStutter) override;
+
+        std::pair<bool, std::vector<uint32_t>> transition_can_produce_to_place(unsigned int t, uint32_t p, ColoredReducer &red, std::vector<uint32_t> already_checked) const;
```

Comment:
Din vector er passed-by-value, hvilket kunne blive dyrt. Hvis du brugte en reference (&) istedet, ville det både være billigere og du behøvede ikke at returnere vectoren heller. Faktisk ville det være endnu bedre at bruge et `std::set` (et set, hvor elementer er sorteret i et træ) i forhold til performance.

Reply:
changed in 1de02419